api_app/api/routes/shared_services.py

```diff
@@ -0,0 +1,89 @@
+from resources import strings
+from .workspaces import save_and_deploy_resource, check_for_etag, construct_location_header
+from azure.cosmos.exceptions import CosmosAccessConditionFailedError
+from .resource_helpers import send_custom_action_message, send_uninstall_message
+from services.authentication import get_current_admin_user, get_current_tre_user_or_tre_admin
+
+
+shared_services_router = APIRouter(dependencies=[Depends(get_current_tre_user_or_tre_admin)])
+
+
+@shared_services_router.get("/shared-services", response_model=SharedServicesInList, name=strings.API_GET_ALL_SHARED_SERVICES, dependencies=[Depends(get_current_tre_user_or_tre_admin)])
+async def retrieve_shared_services(shared_services_repo=Depends(get_repository(SharedServiceRepository))) -> SharedServicesInList:
+    shared_services = shared_services_repo.get_active_shared_services()
+    return SharedServicesInList(sharedServices=shared_services)
+
+
+@shared_services_router.get("/shared-services/{shared_service_id}", response_model=SharedServiceInResponse, name=strings.API_GET_SHARED_SERVICE_BY_ID, dependencies=[Depends(get_current_admin_user), Depends(get_shared_service_by_id_from_path)])
+async def retrieve_shared_service_by_id(shared_service=Depends(get_shared_service_by_id_from_path)) -> SharedServiceInResponse:
+    return SharedServiceInResponse(sharedService=shared_service)
+
+
+@shared_services_router.post("/shared-services", status_code=status.HTTP_202_ACCEPTED, response_model=OperationInResponse, name=strings.API_CREATE_SHARED_SERVICE, dependencies=[Depends(get_current_admin_user)])
+async def create_shared_service(response: Response, shared_service_input: SharedServiceInCreate, shared_services_repo=Depends(get_repository(SharedServiceRepository)), operations_repo=Depends(get_repository(OperationRepository))) -> OperationInResponse:
+    try:
+        shared_service = shared_services_repo.create_shared_service_item(shared_service_input)
+    except (ValidationError, ValueError) as e:
+        logging.error(f"Failed create shared service model instance: {e}")
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
+
+    operation = await save_and_deploy_resource(shared_service, shared_services_repo, operations_repo)
+    response.headers["Location"] = construct_location_header(operation)
+
+    return OperationInResponse(operation=operation)
+
+
+@shared_services_router.patch("/shared-services/{shared_service_id}", response_model=SharedServiceInResponse, name=strings.API_UPDATE_SHARED_SERVICE, dependencies=[Depends(get_current_admin_user), Depends(get_shared_service_by_id_from_path)])
+async def patch_shared_service(shared_service_patch: ResourcePatch, shared_service_repo=Depends(get_repository(SharedServiceRepository)), shared_service=Depends(get_shared_service_by_id_from_path), resource_template_repo=Depends(get_repository(ResourceTemplateRepository)), etag: str = Header(None)) -> SharedServiceInResponse:
+    check_for_etag(etag)
```

Comment:
my 'Resource updates' merge has just gone in underneath you - I wonder if now would be a good chance to copy that logic to handle the updates of shared services in the `PATCH` (see workspace patch route in latest main) ? 'should' only be a few lines change in this method...

Reply:
That sounds like a good idea, will look into it now...